pkg/controller/route/routecontroller.go

```diff
@@ -33,79 +44,217 @@ type RouteController struct {
+	}
+	// if we have no route for node, create one
+	if node != nil && r == nil {
+		glog.Infof("No route for Node %v, create one", key)
+		shouldCreate = true
+	}
+	// check if the route for this node has the correct CIDR, if not, remove the old route and create a new one
+	if node != nil && r != nil && r.DestinationCIDR != node.Spec.PodCIDR {
+		glog.Infof("The route for %v has the incorrect CIDR, remove it and create a new one", key)
+		shouldDelete = true
+		shouldCreate = true
+	}
+
+	if shouldDelete {
+		if err := rc.routes.DeleteRoute(rc.clusterName, r); err != nil {
+			glog.Errorf("Could not delete route %s %s: %v", r.Name, r.DestinationCIDR, err)
+		}
+	}
+
+	if shouldCreate {
+		route := &cloudprovider.Route{
+			TargetInstance:  node.Name,
+			DestinationCIDR: node.Spec.PodCIDR,
+		}
+		nameHint := string(node.UID)
+		if err := rc.routes.CreateRoute(rc.clusterName, nameHint, route); err != nil {
+			glog.Errorf("Could not create route %s %s: %v", nameHint, route.DestinationCIDR, err)
+		}
+	}
+
+	return nil
 }
 
-func (rc *RouteController) reconcile(nodes []api.Node, routes []*cloudprovider.Route) error {
+// checkLeftoverRoutes lists all currently existing routes and detect routes
+// that exist with no corresponding nodes; these routes need to be deleted.
+// We only need to do this once on startup, because in steady-state these are
+// detected (but some stragglers could have been left behind if the route controller
+// reboots).
+func (rc *RouteController) checkLeftoverRoutes() {
+
+	// routeMap maps routeTargetInstance->route
+	routeMap := map[string]*cloudprovider.Route{}
 	// nodeCIDRs maps nodeName->nodeCIDR
 	nodeCIDRs := make(map[string]string)
-	// routeMap maps routeTargetInstance->route
-	routeMap := make(map[string]*cloudprovider.Route)
+
+	routes, err := rc.routes.ListRoutes(rc.clusterName)
+	if err != nil {
+		glog.Errorf("Unable to list routes (%v); orphaned routes will not be cleaned up. (They're pretty harmless, but you can restart this component if you want another attempt made.)", err)
+		return
+	}
 	for _, route := range routes {
 		routeMap[route.TargetInstance] = route
 	}
-	for _, node := range nodes {
-		// Check if we have a route for this node w/ the correct CIDR.
-		r := routeMap[node.Name]
-		if r == nil || r.DestinationCIDR != node.Spec.PodCIDR {
-			// If not, create the route.
-			route := &cloudprovider.Route{
-				TargetInstance:  node.Name,
-				DestinationCIDR: node.Spec.PodCIDR,
-			}
-			nameHint := string(node.UID)
-			go func(nameHint string, route *cloudprovider.Route) {
-				if err := rc.routes.CreateRoute(rc.clusterName, nameHint, route); err != nil {
-					glog.Errorf("Could not create route %s %s: %v", nameHint, route.DestinationCIDR, err)
-				}
-			}(nameHint, route)
-		}
+
+	nodeList, err := rc.nodeStore.List()
+	if err != nil {
+		glog.Errorf("error listing nodes: %v", err)
+		return
+	}
+	for _, node := range nodeList.Items {
 		nodeCIDRs[node.Name] = node.Spec.PodCIDR
```

Comment:
Some nodes might not have PodCIDR set. I think we can skip adding them to the map.